File: app/src/main/java/com/boycoder/todo/MainActivity.java
package com.boycoder.todo;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.ViewGroup;

import androidx.activity.EdgeToEdge;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

public class MainActivity extends AppCompatActivity {
    private TaskViewModel taskViewModel;
    private TaskAdapter taskAdapter;

    private final ActivityResultLauncher<Intent> taskDetailLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK && result.getData() != null) {
                    Intent data = result.getData();
                    String title = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_TITLE);
                    String description = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_DESCRIPTION);
                    String priority = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_PRIORITY);
                    long dueDateValue = data.getLongExtra(TaskDetailActivity.EXTRA_TASK_DUE_DATE, -1);
                    Long dueDate = dueDateValue == -1 ? null : dueDateValue;

                    if (data.hasExtra(TaskDetailActivity.EXTRA_TASK_ID)) {
                        String id = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_ID);
                        
                        // Anti-pattern for Lesson 1: Finding an item with a manual loop
                        // This will be refactored using Kotlin's 'find' or 'firstOrNull'
                        Task updatedTask = null;
                        for (Task t : taskViewModel.getTasks().getValue()) {
                            if (t.getId().equals(id)) {
                                updatedTask = t;
                                break;
                            }
                        }
                        
                        if (updatedTask != null) {
                            // Anti-pattern for Lesson 1: Sequential setters instead of 'apply' or data class 'copy'
                            updatedTask.setTitle(title);
                            updatedTask.setDescription(description);
                            updatedTask.setPriority(priority);
                            updatedTask.setDueDate(dueDate);
                            taskViewModel.updateTask(updatedTask);
                        }
                    } else {
                        // Anti-pattern for Lesson 1: Sequential setters instead of 'apply'
                        Task newTask = new Task(title, description);
                        newTask.setPriority(priority);
                        newTask.setDueDate(dueDate);
                        taskViewModel.addTask(newTask);
                    }
                }
            }
    );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        EdgeToEdge.enable(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        
        RecyclerView recyclerView = findViewById(R.id.recycler_view_tasks);
        FloatingActionButton fab = findViewById(R.id.fab_add_task);
        android.view.View header = findViewById(R.id.text_header);
        
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            
            // Apply status bar inset as top padding to header so it moves down naturally
            header.setPadding(header.getPaddingLeft(), systemBars.top, header.getPaddingRight(), header.getPaddingBottom());
            
            // For Recycler View, apply nav bar inset as bottom padding so items scroll behind nav bar
            int recyclerPaddingBottomPx = (int) (88 * getResources().getDisplayMetrics().density);
            recyclerView.setPadding(
                recyclerView.getPaddingLeft(),
                recyclerView.getPaddingTop(),
                recyclerView.getPaddingRight(),
                systemBars.bottom + recyclerPaddingBottomPx
            );
            
            // Move FAB up by setting bottom margin to include system bar height + original 32dp
            int fabMarginBottomPx = (int) (32 * getResources().getDisplayMetrics().density);
            ViewGroup.MarginLayoutParams fabParams = (ViewGroup.MarginLayoutParams) fab.getLayoutParams();
            fabParams.bottomMargin = systemBars.bottom + fabMarginBottomPx;
            fab.setLayoutParams(fabParams);
            
            return insets;
        });

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        taskAdapter = new TaskAdapter(
            (task, isCompleted) -> {
                task.setCompleted(isCompleted);
                taskViewModel.updateTask(task);
            },
            task -> {
                Intent intent = new Intent(MainActivity.this, TaskDetailActivity.class);
                intent.putExtra(TaskDetailActivity.EXTRA_TASK_ID, task.getId());
                intent.putExtra(TaskDetailActivity.EXTRA_TASK_TITLE, task.getTitle());
                intent.putExtra(TaskDetailActivity.EXTRA_TASK_DESCRIPTION, task.getDescription());
                if (task.getPriority() != null) {
                    intent.putExtra(TaskDetailActivity.EXTRA_TASK_PRIORITY, task.getPriority());
                }
                if (task.getDueDate() != null) {
                    intent.putExtra(TaskDetailActivity.EXTRA_TASK_DUE_DATE, task.getDueDate());
                }
                taskDetailLauncher.launch(intent);
            }
        );
        recyclerView.setAdapter(taskAdapter);

        taskViewModel = new ViewModelProvider(this).get(TaskViewModel.class);

        taskViewModel.getTasks().observe(this, tasks -> {
            taskAdapter.setTasks(tasks);
            
            // Update the header with the active task count
            int activeCount = taskViewModel.getActiveTaskCount();
            if (header instanceof android.widget.TextView) {
                ((android.widget.TextView) header).setText("Tasks (" + activeCount + " left)");
            }
        });

        fab.setOnClickListener(view -> {
            Intent intent = new Intent(MainActivity.this, TaskDetailActivity.class);
            taskDetailLauncher.launch(intent);
        });

        if (taskViewModel.getTasks().getValue() == null || taskViewModel.getTasks().getValue().isEmpty()) {
            // These verbose sequential setter calls are intentional teaching material.
            // They will be refactored using Kotlin's `apply` scope function.
            
            Task t1 = new Task("Doctor appointment", "Annual physical checkup");
            t1.setPriority("High");
            t1.setDueDate(System.currentTimeMillis() + 86400000); // Tomorrow
            taskViewModel.addTask(t1);

            Task t2 = new Task("Buy groceries", "Milk, Eggs, Bread");
            t2.setPriority("Medium");
            taskViewModel.addTask(t2);

            Task t3 = new Task("Clean the garage", null); // Intentional null description
            taskViewModel.addTask(t3);

            Task t4 = new Task("Read a book", "Finish chapter 3");
            t4.setPriority("Low");
            taskViewModel.addTask(t4);

            Task t5 = new Task("Call mom", "Wish her happy birthday");
            taskViewModel.addTask(t5);
        }
    }
}

File: app/src/main/java/com/boycoder/todo/MainActivity.kt
package com.boycoder.todo

import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.CircleCrop
import java.util.concurrent.TimeUnit
import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText
import com.google.android.material.floatingactionbutton.FloatingActionButton

class MainActivity : AppCompatActivity() {
    private val taskViewModel: TaskViewModel by viewModels()
    private lateinit var taskAdapter: TaskAdapter

    private val taskDetailLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK && result.data != null) {
            val data = result.data!!
            val title = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_TITLE) ?: return@registerForActivityResult
            val description = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_DESCRIPTION)
            val priority = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_PRIORITY)
            val dueDateValue = data.getLongExtra(TaskDetailActivity.EXTRA_TASK_DUE_DATE, -1)
            val dueDate = if (dueDateValue == -1L) null else dueDateValue

            if (data.hasExtra(TaskDetailActivity.EXTRA_TASK_ID)) {
                val id = data.getStringExtra(TaskDetailActivity.EXTRA_TASK_ID)

                // Refactored using Kotlin's 'find'
                taskViewModel.tasks.value?.find { it.id == id }?.let { existingTask ->
                    // Refactored using 'copy' from data class, removing mutable setters on original object!
                    val updatedTask = existingTask.copy(
                        title = title,
                        description = description,
                        priority = priority,
                        dueDate = dueDate
                    )
                    taskViewModel.updateTask(updatedTask)
                }
            } else {
                // Refactored using 'apply'
                val newTask = Task(title, description).apply {
                    this.priority = priority
                    this.dueDate = dueDate
                }
                taskViewModel.addTask(newTask)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val recyclerView: RecyclerView = findViewById(R.id.recycler_view_tasks)
        val fab: FloatingActionButton = findViewById(R.id.fab_add_task)
        val searchEdit: EditText = findViewById(R.id.edit_search)
        val userProfileLayout: android.view.View = findViewById(R.id.layout_user_profile)
        val usernameText: TextView = findViewById(R.id.text_username)
        val avatarImage: ImageView = findViewById(R.id.image_avatar)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { _, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())

            userProfileLayout.setPadding(
                userProfileLayout.paddingLeft,
                systemBars.top,
                userProfileLayout.paddingRight,
                userProfileLayout.paddingBottom
            )

            val recyclerPaddingBottomPx = (88 * resources.displayMetrics.density).toInt()
            recyclerView.setPadding(
                recyclerView.paddingLeft,
                recyclerView.paddingTop,
                recyclerView.paddingRight,
                systemBars.bottom + recyclerPaddingBottomPx
            )

            val fabMarginBottomPx = (32 * resources.displayMetrics.density).toInt()
            val fabParams = fab.layoutParams as ViewGroup.MarginLayoutParams
            fabParams.bottomMargin = systemBars.bottom + fabMarginBottomPx
            fab.layoutParams = fabParams

            insets
        }

        recyclerView.layoutManager = LinearLayoutManager(this)

        taskAdapter = TaskAdapter(
            onTaskStateChanged = { task, isCompleted ->
                taskViewModel.updateTask(task.copy(isCompleted = isCompleted))
            },
            onTaskClicked = { task ->
                val intent = Intent(this@MainActivity, TaskDetailActivity::class.java).apply {
                    putExtra(TaskDetailActivity.EXTRA_TASK_ID, task.id)
                    putExtra(TaskDetailActivity.EXTRA_TASK_TITLE, task.title)
                    putExtra(TaskDetailActivity.EXTRA_TASK_DESCRIPTION, task.description)
                    task.priority?.let { putExtra(TaskDetailActivity.EXTRA_TASK_PRIORITY, it) }
                    task.dueDate?.let { putExtra(TaskDetailActivity.EXTRA_TASK_DUE_DATE, it) }
                }
                taskDetailLauncher.launch(intent)
            }
        )
        recyclerView.adapter = taskAdapter

        taskViewModel.user.observe(this) { user ->
            if (user != null) {
                usernameText.text = user.username
                avatarImage.loadAvatar(user.avatarUrl)
            } else {
                usernameText.text = "Loading user..."
            }
        }

        taskViewModel.tasks.observe(this) { tasks ->
            taskAdapter.setTasks(tasks)

            val activeCount = taskViewModel.getActiveTaskCount()
            val header: TextView = findViewById(R.id.text_header)
            header.text = "Tasks ($activeCount left)"
        }

        fab.setOnClickListener {
            val intent = Intent(this@MainActivity, TaskDetailActivity::class.java)
            taskDetailLauncher.launch(intent)
        }

        // 原始回调地狱 (反面教材)
        searchEdit.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                val query = s?.toString() ?: ""
                taskViewModel.searchTasks(query)
            }
        })

        // Fetch initial data from the network sequentially
        taskViewModel.loadDashboardData()
    }
}
